### PR TITLE
CMake build configuration for C++ library

### PR DESCRIPTION
**Pull request by christianrauch**
This PR fixes the examples to include the GoPiGo3 header instead of the source file, cleans up `GoPiGo3.h` to only contain public functions used by examples, and adds a CMake build configuration and a ROS package definition.

Commits:
- renames example files from *.c to *.cpp
- moves function implementations to `GoPiGo3.cpp`
- moves internal function definitions (not used by examples) to `GoPiGo3.cpp`
- includes `GoPiGo3.h` in all examples and compiles with `GoPiGo3.cpp`
- adds a CMake build configuration for the library and examples
- adds a ROS package definition

### CODE REVIEW

File: Software/C/GoPiGo3.cpp [hunk 1]
```diff
@@ -9,7 +9,68 @@
...
+#include <stdexcept>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <linux/spi/spidev.h>
+#include <stdio.h>            // for printf
+#include <string.h>           // for strstr
+#include <sys/time.h>         // for clock_gettime
+#include <unistd.h>
+
+#define FIRMWARE_VERSION_REQUIRED "0.3." // Firmware version needs to start with this
```
Comment (mattallen37) [inline Software/C/GoPiGo3.cpp]: This definition should be left in the .h file. Same applies to the following definitions on lines 24 and 26.

File: Software/C/GoPiGo3.cpp [hunk 2]
```diff
@@ -9,7 +9,68 @@
...
+
+#define SPI_TARGET_SPEED 500000 // SPI target speed of 500kbps
+
+#define SPIDEV_FILE_NAME "/dev/spidev0.1" // File name of SPI
+
+int spi_file_handle = -1;                    // SPI file handle
```
Comment (mattallen37) [inline Software/C/GoPiGo3.cpp]: While this and the following lines (28 through 73) are used for the GoPiGo3, they are not specific to the GoPiGo3, which is why I intentionally left them in the .h file.

**Comment by christianrauch** [comment]
I've undone most of the changes related to the SPI functions. I.e. the functions are now in the `GoPiGo3.cpp` again.

**Comment by mattallen37** [comment]
Thanks for taking a look at it, we're happy to have you looking over the code.

**Pull request merged**